### ISSUE TITLE
Cross-package false-positives

### ISSUE THREAD

**Issue by NorfairKing** [ocharles/weeder#147]
It looks like weedle doesn't recognise cross-package edges.

**Comment by ocharles** [comment]
Can you provide a bit more information? Weeder is used at CircuitHub with a `cabal.project` with over 100 components and works as expected. Are you provided a consistent set of `.hie` files, or have I misunderstood what you mean by cross-package?

**Comment by NorfairKing** [comment]
@ocharles Let me put together a repro and get back to you!

**Comment by NorfairKing** [comment]
@ocharles Ok I've got an example.
It's when you have packages A and B where the test suite of B depends on the library of A, I think.

You can repro this with:
```
nix build github:Norfairking/smos/weeder#checks.x86_64-linux.weeder-check
```

You'll see this weed (among many atm):

```
weeder-check> src/Smos/Directory/TestUtils.hs:10: withInterestingStore
```

And here are (some of) its appearances:

```
smos-report-gen/test/Smos/Report/FreeSpec.hs
91:          withInterestingStore $ \dc -> do
99:              withInterestingStore $ \dc -> do
109:                  withInterestingStore $ \dc -> do

smos-directory-gen/src/Smos/Directory/TestUtils.hs
9:withInterestingStore :: (DirectorySettings -> IO ()) -> Property
10:withInterestingStore func = forAllValid $ \is -> withDirectorySettings is func
```

**Comment by NorfairKing** [comment]
This is a complete guess, but I think it may have something to do with the directory structure of test suites' source code in nix derivations (?) as they are copied here:

https://github.com/NorfairKing/weeder-nix/blob/aa7fc9ba64c895c33cba0beaf1ca9e21b90324b5/nix/addHieOutput.nix#L27-L33

**Comment by ocharles** [comment]
I would definitely start by looking at your input to weeder. Can you reproduce this problem without Nix? If not, it would suggest something is up with the Nix invocation. We use Haskell.nix for our builds, and then collect all `.hie` files separately at the end:

```nix
  weeder =
    pkgs.runCommand
      "weeder"
      {
        buildInputs = [ pkgs.glibcLocales pkgs.weeder ];
        allHieFiles = pkgs.linkFarm "all-hie-files"
          (
            with pkgs.lib;
            concatMap
              (
                package:
                let
                  pkg-name = package.identifier.name;

                  lib-hies =
                    optional (package.components ? library)
                      {
                        name = "${pkg-name}-library";
                        path = package.components.library.hie;
                      };

                  exe-hies =
                    concatMap
                      (
                        exe: optional (package.components.exes."${exe}" ? hie) {
                          name = "${pkg-name}-exe-${exe}";
                          path = package.components.exes."${exe}".hie;
                        }
                      )
                      (attrNames package.components.exes);

                  test-hies = concatMap
                    (
                      test-name:
                      let
                        test = package.components.tests."${test-name}";

                        is-doctest =
                          # doctest tests build _all_ components of a package.
                          # The GHC id of these packages will be different,
                          # which means that when we run weeder, all this code
                          # will be uncalled.  These are false positives, so
                          # we don't include hie files from anything that
                          # depends on `doctest`.
                          any
                            (x: x.identifier.name or "" == "doctest")
                            test.config.depends;
                      in
                      optional (!is-doctest) {
                        name = "${pkg-name}-test-${test-name}";
                        path = test.hie;
                      }
                    )
                    (attrNames package.components.tests);
                in
                lib-hies ++ exe-hies ++ test-hies
              )
              (
                builtins.attrValues
                  (
                    pkgs.haskell-nix.haskellLib.selectProjectPackages
                      pkgs.circuithub-haskell-nix.hsPkgs
                  )
              )
          );
      } ''
      export LC_ALL=en_US.UTF-8 LANG=en_US.UTF-8 LANGUAGE=en_US.UTF-8
      cd $allHieFiles
      weeder --config ${../../weeder.toml} -j8
      touch $out
    '';
```

**Comment by NorfairKing** [comment]
@ocharles Thanks for the input!
I just failed to reproduce this problem without nix (using stack) so it looks like the problem is indeed with the way the hie files are collected.

Could it be that `weeder` expects a certain directory structure for the `.hie` files? Could we get rid of that requirement?

**Comment by ocharles** [comment]
I'm pretty sure, but I can't rule out there being a bug. Can you try:

* Running with `-j1`
* Running with only one `--hie-directory` flag?
* Running `weeder` outside Nix, but point to one or more directories of `.hie` files generated by Nix

It could also be that the way your building your `.hie` files in Nix are creating different packages ids for everything. I think you might be able to debug that with the `dump` command in `hiedb` (I think it might be `dump` or `dump-ast`).

**Comment by NorfairKing** [comment]
@ocharles Alright, some more debugging later.
First of all: Doh, my comment about thread safety was silly because I'd only been running with `-j1` so far (the default, right?).

I just ran weeder outside of nix with multiple `--hie-directory` options like this:
```
weeder --hie-directory ./smos-directory-gen/.stack-work/dist/x86_64-linux/ghc-9.4.8/build --hie-directory ./smos-report-gen/.stack-work/dist/x86_64-linux/ghc-9.4.8/build 
```
These directories have the same funky structure as the nix-build does, and weeder doesn't show the false-positives, so we know that it's not (just) about the directory structure and not about the multiple flags.

I'll continue digging!

**Comment by NorfairKing** [comment]
Thank you very much for your note on package ids.
I was making the same mistake that I spent ages debugging for [dekking]( https://github.com/NorfairKing/dekking).
I was adding a `.hie` output to packages, but did not put that overriden package back into the `haskellPackages` via an override.
As a result, packages looked like this:

```
dirs:
- addHieDir A -> B
- addHieDir B       
```
In other words, the overriden version of A still depended on the un-overridden version of B, causing the mismatched package ids that you described (I think).

I've got some more refactoring to do, but this issue is fixed/didn't exist in the firs place so I'll close it.
Thank you for the help and the patience!
